trace-utils/src/msgpack_decoder/v04/decoder/mod.rs

```diff
@@ -99,54 +105,81 @@ fn read_string_ref(buf: &[u8]) -> Result<(&str, &[u8]), DecodeError> {
 }
 
 #[inline]
-fn read_string(buf: &mut &[u8]) -> Result<String, DecodeError> {
-    let (str_ref, remaining_buf) = read_string_ref(buf)?;
-    *buf = remaining_buf;
-    Ok(str_ref.to_string())
-}
+// Safety: read_string_ref checks utf8 validity, so we don't do it again when creating the
+// NoAllocStrings.
+fn read_str_map_to_no_alloc_strings(
+    buf_wrapper: &BufferWrapper,
+    buf: &mut &[u8],
+) -> Result<HashMap<NoAllocString, NoAllocString>, DecodeError> {
+    let len = decode::read_map_len(buf)
+        .map_err(|_| DecodeError::InvalidFormat("Unable to get map len for str map".to_owned()))?;
 
-#[inline]
-fn read_str_pair(buf: &mut &[u8]) -> Result<(String, String), DecodeError> {
-    let k = read_string(buf)?;
-    let v = read_string(buf)?;
+    let mut map = HashMap::with_capacity(len.try_into().expect("Unable to cast map len to usize"));
+    for _ in 0..len {
+        let (key, next) = read_string_ref(buf)?;
+        *buf = next;
 
-    Ok((k, v))
+        let (val, next) = read_string_ref(buf)?;
+        *buf = next;
+
+        let key = buf_wrapper.create_no_alloc_string_unchecked(key.as_bytes());
```

Comment:
I think it would be better to have a function for calling `read_string_ref` and converting the slice with `unchecked` this way we have only one function dealing with potentially unsafe code.